### ISSUE TITLE
false `MoreSpecificReturnType` when returning a `numeric-string` for `non-empty-string` return.

### ISSUE THREAD

**Issue by azjezz** [vimeo/psalm#5141]
after asserting that `numeric-string` is `non-empty-string`, psalm doesn't seem to convert the variable type internally.

https://psalm.dev/r/ad022887b0

**Comment by orklah** [comment]
I'm not sure it should.

This is insanely specific so I don't think it's worth creating a new type `non-empty-numeric-string` (which is just any numeric-string except 0)

Changing the type to non-empty-string will prevent any numeric operation on this variable and create another set of false-positives.

**Comment by azjezz** [comment]
`int(0)` when cast to a string is `string("0")`, so it should also be considered a non-empty-string.

IMO, all `numeric-strings` should be considered `non-empty-string` ( unless there's another edge case? )

**Comment by orklah** [comment]
there is no "also a `non-empty-string`" notion in Psalm. It's either a `non-empty-string` OR a `numeric-string`.

And if you transform the `numeric-string` into a `non-empty-string`, you'll then get warnings on any numeric operation (for example, bc_add won't accept `non-empty-string`)

Except `(string) 0` and `(string) 0.0`, I don't think there are other cases of empty numeric-string but it's enough to make `numeric-string` potentially empty...

**Comment by azjezz** [comment]
> there is no "also a non-empty-string" notion in Psalm. It's either a non-empty-string OR a numeric-string.

by "also", I mean that `numeric-string` *should* be a subtype of `non-empty-string`, just like `non-empty-string` is a subtype of `string`.

> And if you transform the numeric-string into a non-empty-string, you'll then get warnings on any numeric operation (for example, bc_add won't accept non-empty-string)

That is not an issue in the case described above, as we want to return a `non-empty-string`, regardless if it's numeric or not.

> Except (string) 0 and (string) 0.0, I don't think there are other cases of empty numeric-string but it's enough to make numeric-string potentially empty...

`(string) 0` does not result in an empty string tho ... https://3v4l.org/Qqvn8 ( string `"0"` is not considered "empty" )

**Comment by weirdan** [comment]
> string "0" is not considered "empty"

It is: https://3v4l.org/pF4sm

**Comment by orklah** [comment]
>  string "0" is not considered "empty"
Oh I see, sorry, I should have been more clear. `non-empty-string` does not mean that the string is literaly empty. It means calling `empty()` on it will return true.

**Comment by azjezz** [comment]
> It is: https://3v4l.org/pF4sm

> non-empty-string does not mean that the string is literally empty. It means calling empty() on it will return false.

I don't think that's the case in psalm, as the following is okay:

```

/**
 * @return null|non-empty-string
 */
function to_non_empty_string(mixed $value): ?string
{
    if (is_string($value) && '' !== $value) {
        return $value;
    }

    return null;
}
```

if `"0"` is considered empty in psalm, then the above code *shall not pass* as we might be returning `"0"`.

**Comment by orklah** [comment]
This is a bug `'' !== $value` shouldn't rule out empty-string (`!=` probably might).

The definition of empty is tightly linked to empty():
https://psalm.dev/r/3abd6c9058

And more generally with the falsy notion:
https://psalm.dev/r/f497fc7132

**Comment by azjezz** [comment]
> The definition of empty is tightly linked to empty():

This is not mentioned anywhere in the documentation tho https://psalm.dev/docs/annotating_code/type_syntax/scalar_types/#lowercase-string-non-empty-string-non-empty-lowercase-string

> This is a bug '' !== $value shouldn't rule out empty-string (!= probably might).

actually, both should, as both do rule it out.

see: https://3v4l.org/WkaOO


IMHO, there's two bugs here:

1. numeric-string is never `string(0) ""`, so it should be a subtype of `non-empty-string`
2. psalm assumes when `empty($str) === false`, that `$str` is `non-empty-string`, which is true, however, psalm also assumes that if `empty($str) === true` that `$str` cannot be `non-empty-string`, and that's false, as `string(1) "0"` is technically a `non-empty-string` ( typed checked: https://psalm.dev/r/161af8f453 ), but can be considered "empty" in php terms ( not psalm terms ).

**Comment by orklah** [comment]
Well, if `non-empty-string` was only `strong(0) ""`, it wouldn't need a dedicated type in Psalm. It would just be a `TLiteralString` with value `""`.

Also, the non-empty-string would suddenly become nearly useless because it wouldn't help at all with `if($myString)` nor `if(empty($myString))` situations.

But it seems Webmozart\Assert use your definition for stringNotEmpty: https://github.com/webmozarts/assert/blob/master/src/Assert.php#L59

**Comment by azjezz** [comment]
> But it seems Webmozart\Assert use your definition for stringNotEmpty: https://github.com/webmozarts/assert/blob/master/src/Assert.php#L59

[`azjezz/psl`](https://github.com/azjezz/psl) also uses the same definition for what `non-empty-string` is ( e.g: Str\is_empty, Type\non_empty_string, and multiple other places )

and most of my projects rely on this behavior ( i.e: non-empty-string = anything but `string(0) ""` ), 

> useless because it wouldn't help at all with if($myString) nor if(empty($myString)) situations.

that is what `@param non-empty-string $bar` is supposed to protect against, i don't have to check if($bar) or if(empty($bar)) if i can enforce that it is a non empty string ( either in ci using psalm, or at runtime using webmozarts/assert/ or zjezz/psl ).

**Comment by muglug** [comment]
As to the `non-empty-string` thing, I realise it's very confusing. I thinking about it for a while I believe `non-empty-string` should match all strings except `""`. I think there's _also_ scope for having a subtype of `non-empty-string` called `non-falsy-string` that prohibits `"0"` and `""` and is the result of `is_string($s) && $s`

**Comment by orklah** [comment]
According to this definition, numeric-string should indeed be a supertype of numeric-string, also empty should not transform numeric-string into string

**Comment by azjezz** [comment]
```
                                           string
                                               |
                                               |
                              non-empty-string ( not string(0) "" )
                                               |
                  |-----------------------------------------------------------------------------------------|
                  |                                                                                         |
non-falsy-string ( not string(0) "", not string(1) "0" )                        numeric-string ( usually result of (string) $num )

```

**Comment by muglug** [comment]
OK, so that's done in a0420fb70.

I agree it's weird, but PHP is weird. I've recently been caught out by `if (!$type) {...}` where `$type` was `0`.

**Comment by orklah** [comment]
well, I think the core of this issue was to consider numeric-string as non-empty-string. 

The example still fails: https://psalm.dev/r/ad022887b0

**Comment by muglug** [comment]
@orklah I know, which is why I hadn't yet closed the ticket. But now I have 😄

**Comment by orklah** [comment]
Thanks then :)

I posted two comments on the first commit, I don't know if you see them

**Comment by azjezz** [comment]
Thank you for the fix @muglug 🎉
